File: comparison/simplified.R
# Load libs, data ---------------------------------------------------------
# Specify log likelihood func (wrapper on rtdists)-------------------------
# Specify the parameters and priors ---------------------------------------
pars <- c("b1", "b2", "b3", "A", "v1", "v2", "t0")
priors <- list(
  theta_mu = rep(0, length(pars)),
  theta_sig = diag(rep(1, length(pars)))
)
# Create the Particle Metropolis within Gibbs sampler object --------------
sampler <- pmwgs(
  data = forstmann,
  pars = pars,
  ll_func = lba_loglike,
  prior = priors
)
# Start points ------------------------------------------------------------
start_points <- list(
  mu = c(.2, .2, .2, .4, .3, 1.3, -2),
  sig2 = diag(rep(.01, length(pars)))
)
# No setup yet for number iterations etc - all in run_stage ---------------







# Initialise the sampler --------------------------------------------------
sampler <- init(sampler, theta_mu = start_points$mu,
                theta_sig = start_points$sig2)
# Copy of init code -------------------------------------------------------
init.pmwgs <- function(x, theta_mu=NULL, theta_sig=NULL, display_progress=TRUE, ...) {
  # Setup num particles, creating output array alpha etc
  for (s in 1:x$n_subjects) {
    particles <- mvtnorm::rmvnorm(n_particles, theta_mu, theta_sig)
    #Using rtdists so not needed to manipulate particles into diff format
    lw <- apply(particles, 1, x$ll_func, data = x$data[x$data$subject == x$subjects[s], ])
    # Our loglike function tests for impossible numbers etc not needed here
    weight <- exp(lw - max(lw))

    # We don't divide by sum because sample function can take weights not summing to 1
    # Additional check for negative weights, etc
    idx <- sample(x = n_particles, size = 1, prob = weight)
    alpha[, s] <- particles[idx, ]
    likelihoods[s] <- lw[idx]
  }
  # Set stores in x (pmwgs object) with alpha, likelihoods etc
}
# No more setup needed - ready to run stages-------------------------------
# Run_stage function, meat of processing ----------------------------------
burned <- run_stage(sampler, stage = "burn")
run_stage.pmwgs <- function(x, stage, iter = 1000, particles = 1000, display_progress = TRUE, n_cores = 1, ...) {
# Select stage, check args build sample storage----------------------------
# If stage is sample, try to create_efficient proposal args ---------------
  for (i in 1:iter) {
# Call Function for gibbs step, plus function contents --------------------
      pars <- new_group_pars(store, x),
new_group_pars <- function(samples, sampler) {
# Get last sample, hyper pars ---------------------------------------------
  var_mu <- MASS::ginv(sampler$n_subjects * last$gvi + sampler$prior$theta_sig_inv)
  mean_mu <- as.vector(var_mu %*% (last$gvi %*% apply(last$sm, 1, sum)))
  chol_var_mu <- t(chol(var_mu)) # t() because I want lower triangle.
  gm <- mvtnorm::rmvnorm(1, mean_mu, chol_var_mu %*% t(chol_var_mu))[1, ]
  # k_half calculated earlier
  theta_temp <- last$sm - gm
  cov_temp <- (theta_temp) %*% (t(theta_temp))


  
  B_half <- 2 * hyper$v_half * diag(1 / hyper$a_half) + cov_temp #nolint
  gv <- MCMCpack::riwish(hyper$k_half, B_half) # New sample for group variance
  gvi <- MASS::ginv(gv)

  # Sample new mixing weights.
  a_half <- 1 / stats::rgamma(
    n = sampler$n_pars,
    shape = hyper$v_shape,
    scale = 1 / (hyper$v_half + diag(gvi) + hyper$A_half)
  )
#  return calculated values
}
# Set arguments and functions for getting new samples---------------------
# Contents of create_efficient (called earlier) repro'ed for comparison ----
create_efficient <- function(x) {
  # Create storage for conditional pars, by subject run conditional_parms
}
conditional_parms <- function(s, samples) {
  all_samples <- rbind(
    samples$alpha[, s, ],
    samples$theta_mu[, ],
    pts2_unwound
  ) # pts_unwound is:
  #y <- t(chol(var_matrix))  # take cholesky factor (top right), transpose
  #diag(y) <- log(diag(y))   # log the diagonal
  #y[lower.tri(y, diag = TRUE)]  
# Create sample mean sigma for joint random effect and pars ---------------------
  mu_tilde <- apply(all_samples, 1, mean)
  sigma_tilde <- stats::var(t(all_samples))
  #var will computer covariance matrix, we don't correct for non pos definite
# R code calculates conidtional mean/var here, code later for comparison)-----
  condmvn <- condMVNorm::condMVN(...)
# Switching num particles done at higher level----------------------------------------------
# Obtain a new sample -----------------------------------------------------
    tmp <- do.call(apply_fn, fn_args)
new_sample <- function(s, data, num_particles, parameters,
                       efficient_mu = NULL, efficient_sig2 = NULL,
                       mix_ratio = c(0.5, 0.5, 0.0),
                       likelihood_func = NULL,
                       epsilon = 1, subjects = NULL) {
# Already at subject level
# Check args, extract efficient proposal means/last sample pars -----------
# no equivalent here but copied from above---------------------------------
  condmvn <- condMVNorm::condMVN(
    mean = mu_tilde,
    sigma = sigma_tilde,
    dependent.ind = 1:n_par,
    given.ind = (n_par + 1):length(mu_tilde),
    X.given = c(samples$theta_mu[, n_iter],
                unwind(samples$theta_sig[, , n_iter]))
  )
  list(cmeans = condmvn$condMean, cvars = condmvn$condVar)
}
# Create proposals for new particles --------------------------------------
  proposals <- gen_particles(num_particles, mu, sig2, subj_mu, mix_ratio = mix_ratio, prop_mu = e_mu, prop_sig2 = e_sig2, epsilon = epsilon )
  # Except here:
  particle_numbers <- numbers_from_ratio(mix_ratio, num_particles)
  pop_particles <- particle_draws(particle_numbers[1], mu, sig2)
  ind_particles <- particle_draws(particle_numbers[2], particle, sig2 * epsilon)
  eff_particles <- particle_draws(particle_numbers[3], prop_mu, prop_sig2)
  particles <- rbind(pop_particles, ind_particles, eff_particles)
  # Put the current particle in slot 1.
  proposals[1, ] <- subj_mu
# No rearranging necessary -----------------------------------------------

# Again no rearrangin necessary ------------------------------------------
# computing the log density of the LBA given the particles of random effects-----
  lw <- apply(proposals, 1, likelihood_func, data = data[data$subject == subjects[s], ])


  
# Density of random effects proposal given population-level distribution.----
  lp <- mvtnorm::dmvnorm(x = proposals, mean = mu, sigma = sig2, log = TRUE)
  # Density of proposals given proposal distribution.
  prop_density <- mvtnorm::dmvnorm(x = proposals, mean = subj_mu, sigma = sig2)
  # Density of efficient proposals
  if (mix_ratio[3] != 0) {
    eff_density <- mvtnorm::dmvnorm(
      x = proposals,
      mean = e_mu,
      sigma = e_sig2
    )
  } else {
    eff_density <- 0
  }
l
  lm <- log(mix_ratio[1] * exp(lp) +
    (mix_ratio[2] * prop_density) +
    (mix_ratio[3] * eff_density))
  # log of importance weights.
  l <- lw + lp - lm
# No checks here for imaginary number of logw ----------------------------
  weights <- exp(l - max(l))





# Sample winning particle ------------------------------------------------
  idx <- sample(x = num_particles, size = 1, prob = weights)
  winner <- proposals[idx, ]
  attr(winner, "ll") <- lw[idx]
  winner
}





# storing the MCMC draws------------------------------------------------
# Check for number unique values in random effects----------------------
# Save up to the user (RData for example)-------------------------------

File: comparison/simplified.m
%% Load data, no libs, just copied code files----------------------------
%% Log likelihood function in LBA_tpdf, LBA_tcdf files---------------------
%% Parameters unnamed, priors set below-----------------------------------
num_randeffect=7; %the total number of random effects in the LBA model. For Forstmann dataset, we have 7 random effects.
prior_mu_mean=zeros(num_randeffect,1); %the prior for \mu_{\alpha}
prior_mu_sig2=eye(num_randeffect);%the prior for \Sigma_{\alpha}


%% No equivalent of pmwgs option, set hyper parameters here--------------
v_half=2; %the hyperparameters of the prior of \Sigma_{\alpha}
A_half=1; %the hyperparameters of the prior of \Sigma_{\alpha}




%% Initial values for theta_mu/theta_sig2-----------------------------------
param.theta_mu=[0.2695;0.2116;-0.0241;-0.4017;0.2978;1.1230;-1.7128]; %the initial values for parameter \mu
param.theta_sig2=iwishrnd(eye(num_randeffect),20); % the initial values for \Sigma


%% Set up number iters, number burnin etc-----------------------------------
param.sv=1; %set across-trial variability in drift rate to 1 as scaling constant.
param.num_randeffect=7; %total number of random effects (subject-level parameters) in the LBA model. For Forstmann dataset, we have 7 random effects.

num_choice=2; % the number of response options in the task
burn=500; % number of burn in iterations
nit=12000; % we take the last 10000 draws out of 12000 draws
s=100000;% the maximum number of iterations the algorithm will run
%% Initialise the random effects------------------------------------------
[theta_latent]=LBA_MC_v1(data,param,num_subjects,num_trials,num_particles); %obtain initial values of the random effects.

%% Copy of init code-----------------------------------------------------
function [theta_latent]=LBA_MC_v1(data,param,num_subjects,num_trials,num_particles)
% No setup - straight into iterate over subject
parfor j=1:num_subjects
    rnorm=mvnrnd(param.theta_mu',param.theta_sig2,num_particles);
    % Manipulate particles (rnorm) and data into format to run log likelihood function
    lw=real(log(LBA_n1PDF_reparam_real(data_rt_repmat, rnorm_theta_A_kron, rnorm_theta_b_kron, rnorm_theta_v_kron, ones(num_particles*num_trials(j,1),1),rnorm_theta_tau_kron)));
    % Check for real values from LBA function, reshape results
    max_logw=max(real(logw));
    weight=real(exp(logw-max_logw));
    weight=weight./sum(weight);
    % M: Check for neg weights, reshape rnorm_theta etc
    if Nw>0 
        ind=randsample(Nw,1,true,weight);
        theta_latent(j,:)=rnorm_theta(ind,:);
    end
% After looping return theta_latent

%% More setup of vals for gibbs step, storage for conditional step etc --
%% No function for looping, ---------------------------------------------


%% so no checking,-------------------------------------------------------
%% conditional args created later----------------------------------------
while i<=s
%% Gibbs step done inline-------------------------------------------------


%% Last sample/hyper pars already available -----------------------------
    var_mu=inv(num_subjects*inv(param.theta_sig2)+inv(prior_mu_sig2));
    mean_mu=var_mu*(inv(param.theta_sig2)*(sum(theta_latent)'));
    chol_var_mu=chol(var_mu,'lower');
    param.theta_mu=mvnrnd(mean_mu,chol_var_mu*chol_var_mu');
    k_half=v_half+num_randeffect-1+num_subjects;
    cov_temp=zeros(num_randeffect,num_randeffect);
    for j=1:num_subjects
        theta_j=theta_latent(j,:)';
        cov_temp=cov_temp+(theta_j-param.theta_mu')*(theta_j-param.theta_mu')';
    end
    B_half=2*v_half*diag([1./a_half])+cov_temp;
    param.theta_sig2=iwishrnd(B_half,k_half);    
    theta_sig2_inv=inv(param.theta_sig2);

    %sample a_{1},...,a_{7}, see algorithm 3 (2c) of the paper for the full conditional distribution
    for j=1:num_randeffect
        temp_v_half=(v_half+num_randeffect)/2;
        temp_s_half=(v_half*theta_sig2_inv(j,j)+A_half);
        a_half(j,1)=1./random('gam',temp_v_half,1/temp_s_half);
    end
 

%% No equivalent for setting to grab new samples
%% Create efficient (conditional) pars for sampling phase
   % Carlo algorithm
   if sum(count>switch_num)==num_subjects
         for j=1:num_subjects 

             theta=[theta_latent_b1_store(burn:end,j),theta_latent_b2_store(burn:end,j),theta_latent_b3_store(burn:end,j),...
             theta_latent_A_store(burn:end,j),theta_latent_v1_store(burn:end,j),theta_latent_v2_store(burn:end,j),theta_latent_tau_store(burn:end,j),...
             theta_mu_store(burn:end,:),chol_theta_sig2_store1(burn:end,:),chol_theta_sig2_store2(burn:end,:),chol_theta_sig2_store3(burn:end,:),chol_theta_sig2_store4(burn:end,:),chol_theta_sig2_store5(burn:end,:),...
             chol_theta_sig2_store6(burn:end,:),chol_theta_sig2_store7(burn:end,:)];




%% Create covariance matrix and mean for joint random effect and pars----
             covmat_theta(:,:,j)=cov(theta); %computing sample covariance matrix for the joint random effects and parameters \mu_{\alpha} and \Sigma_{\alpha}
             covmat_theta(:,:,j)=topdm(covmat_theta(:,:,j)); %little correction if the covariance matrix for the proposal is not positive definite matrix.
             mean_theta(j,:)=mean(theta); %computing the sample mean for the joint random effects and parameters \mu_{\alpha} and \Sigma_{\alpha}
%% Not done at this point------------------------------------------------

%% Switching num particles etc-------------------------------------------
%% Run conditional monte carlo here--------------------------------------
   [theta_latent]=LBA_CMC_v1(data,param,theta_latent,num_subjects,num_trials,num_particles,mean_theta,covmat_theta,i,count,switch_num);
function [theta_latent,llh]=LBA_CMC_v1(data,param,...
theta_latent,...
num_subjects,num_trials,num_particles,mean_theta,covmat_theta,i,count,switch_num)
%this is the Conditional Monte Carlo algorithm to sample the random effects
%for each subject.
parfor j=1:num_subjects
%% Set scaling parameter, extract last set of random effects (ref par)---        
%% If n.unique good, update mixing pars, num particle to generate and grab conditional mean/var------
% unwind chol_theta_sig2
        xx=[param.theta_mu';chol_theta_sig2_1';chol_theta_sig2_2';chol_theta_sig2_3';...
            chol_theta_sig2_4';chol_theta_sig2_5';chol_theta_sig2_6';chol_theta_sig2_7'];% we need this to compute the mean of the proposal in the sampling stage
        cond_mean=mean_theta(j,1:7)'+covmat_theta(1:7,8:end,j)*((covmat_theta(8:end,8:end,j))\(xx-mean_theta(j,8:end)')); % computing the mean of the proposal in the sampling stage
        cond_var=covmat_theta(1:7,1:7,j)-covmat_theta(1:7,8:end,j)*(covmat_theta(8:end,8:end,j)\covmat_theta(8:end,1:7,j)); % computing the variance of the proposal in the sampling stage
        chol_cond_var=chol(cond_var,'lower');
%condmean = mean_theta + covmat_theta
%
%
%

%% Create proposal particles -------------------------------------------
        rnorm1=cond_mean+chol_cond_var*randn(param.num_randeffect,n1);
        chol_covmat=chol(param.theta_sig2,'lower');
        rnorm2=param.theta_mu'+chol_covmat*randn(param.num_randeffect,n2);
        % Only in 1st two stages (non conditional)
        rnorm1=reference_par'+epsilon.*chol_covmat*randn(7,n1);

        rnorm=[rnorm1,rnorm2];


%% Rearrange arrays ----------------------------------------------------
    rnorm_theta=[rnorm_theta_b1,rnorm_theta_b2,rnorm_theta_b3,rnorm_theta_A,rnorm_theta_v1,rnorm_theta_v2,rnorm_theta_tau];
%% more rearranging to get inn format for :LBA code --------------------
%%computing the log density of the LBA given the particles of random effects---
    lw=real(log(LBA_n1PDF_reparam_real(data_rt_repmat, rnorm_theta_A_kron, rnorm_theta_b_kron, rnorm_theta_v_kron, ones(num_particles*num_trials(j,1),1),rnorm_theta_tau_kron)));
    lw_reshape=reshape(lw,num_trials(j,1),num_particles);
    logw_first=sum(lw_reshape);
    
%% Other densities depending on stage of sampling----------------------------
    if  sum(count>switch_num)==num_subjects
        logw_second=(logmvnpdf(rnorm_theta,param.theta_mu,chol_covmat*chol_covmat'));
        logw_third=log(w_mix.*mvnpdf(rnorm_theta,cond_mean',chol_cond_var*chol_cond_var')+...
            (1-w_mix).*mvnpdf(rnorm_theta,param.theta_mu,chol_covmat*chol_covmat'));
        logw=logw_first'+logw_second'-logw_third;


    else
        logw_second=(logmvnpdf(rnorm_theta,param.theta_mu,chol_covmat*chol_covmat'));
        logw_third=log(w_mix.*mvnpdf(rnorm_theta,reference_par,(epsilon^2).*(chol_covmat*chol_covmat'))+...
            (1-w_mix).*mvnpdf(rnorm_theta,param.theta_mu,chol_covmat*chol_covmat'));
        logw=logw_first'+logw_second'-logw_third;
    end
    
	% mvnpdf(rnorm_theta, ref_par, epsilon^2*(chol_covmat*chol_covmat') 






%%check if there is imaginary number of logw----------------------------
    max_logw=max(real(logw));
    weight=real(exp(logw-max_logw));
    llh_i(j) = max_logw+log(mean(weight)); 
    llh_i(j) = real(llh_i(j)); 	
    weight=weight./sum(weight);
    % More checks for negative weights
%% Sample a winning particle--------------------------------------------
    if Nw>0 
        ind=randsample(Nw,1,true,weight);
        theta_latent(j,:)=rnorm_theta(ind,:);
    end        
    
end
llh=sum(llh_i);
end
   
%% storing the MCMC draws------------------------------------------------
%% Check for number unique values in random effects----------------------
%% Save every N iterations and save final dataset------------------------
